File: linux-client/osm_ride_linux/ui/trend_chart.py
"""Value-vs-time strip, unit-agnostic (used identically for weight and waist since callers
pre-convert to display units, e.g. lb/in) - raw daily points as a thin faded line/dots, with a
bold 7-day rolling average line and filled area on top. Daily weight/waist fluctuates from water,
sodium, timing, etc., so the average is the actionable trend, but showing the raw noise too keeps
a bad single day from misreading as a real regression.

Mirrors app/src/main/java/com/ewaldmire/osmride/ui/bodymetrics/TrendChart.kt.
"""

from __future__ import annotations

import datetime
import math

import cairo
import gi

gi.require_version("Gtk", "4.0")
from gi.repository import Gtk  # noqa: E402

from ..weight.trend import TrendPoint, rolling_average  # noqa: E402


def _short_date(epoch_millis: int) -> str:
    return datetime.datetime.fromtimestamp(epoch_millis / 1000).strftime("%b %-d")


class TrendChart(Gtk.DrawingArea):
    def __init__(self) -> None:
        super().__init__()
        self._points: list[TrendPoint] = []
        self.set_size_request(-1, 120)
        self.set_draw_func(self._on_draw)

    def set_points(self, points: list[TrendPoint]) -> None:
        # points may arrive in any order; the chart reads left-to-right chronologically.
        self._points = sorted(points, key=lambda p: p.epoch_millis)
        self.queue_draw()

    def _on_draw(self, _area: Gtk.DrawingArea, cr, width: int, height: int) -> None:  # noqa: ANN001 - cairo.Context
        cr.set_source_rgb(0.85, 0.85, 0.85)
        cr.rectangle(0, 0, width, height)
        cr.fill()

        if len(self._points) < 2:
            return

        averaged = rolling_average(self._points)
        all_values = [p.value for p in self._points] + [p.value for p in averaged]
        min_value = min(all_values)
        max_value = max(max(all_values), min_value + 0.1)
        min_time = self._points[0].epoch_millis
        max_time = max(self._points[-1].epoch_millis, min_time + 1)

        def x_for(epoch_millis: int) -> float:
            return (epoch_millis - min_time) / (max_time - min_time) * width

        def y_for(value: float) -> float:
            return height - (value - min_value) / (max_value - min_value) * height

        # Raw daily points - thin, faded, no fill, so they read as background noise not signal.
        cr.set_source_rgba(0.1, 0.1, 0.1, 0.35)
        cr.set_line_width(2)
        cr.move_to(x_for(self._points[0].epoch_millis), y_for(self._points[0].value))
        for point in self._points[1:]:
            cr.line_to(x_for(point.epoch_millis), y_for(point.value))
        cr.stroke()
        for point in self._points:
            cr.arc(x_for(point.epoch_millis), y_for(point.value), 3, 0, 2 * math.pi)
            cr.fill()

        # 7-day rolling average - the actual signal, drawn bold with a filled area underneath.
        cr.set_source_rgba(0.13, 0.44, 0.71, 0.5)
        cr.move_to(0, height)
        for point in averaged:
            cr.line_to(x_for(point.epoch_millis), y_for(point.value))
        cr.line_to(width, height)
        cr.close_path()
        cr.fill()

        cr.set_source_rgb(0.13, 0.44, 0.71)
        cr.set_line_width(4)
        cr.move_to(x_for(averaged[0].epoch_millis), y_for(averaged[0].value))
        for point in averaged[1:]:
            cr.line_to(x_for(point.epoch_millis), y_for(point.value))
        cr.stroke()

        # Without these, the line's x-position was scaled by real elapsed time but nothing on
        # screen actually told the viewer that - it just looked like an arbitrary shape.
        cr.select_font_face("sans-serif", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_NORMAL)
        cr.set_font_size(12)
        cr.set_source_rgb(0.1, 0.1, 0.1)
        start_label = _short_date(min_time)
        end_label = _short_date(max_time)
        cr.move_to(4, height - 6)
        cr.show_text(start_label)
        end_extents = cr.text_extents(end_label)
        cr.move_to(width - end_extents.width - 4, height - 6)
        cr.show_text(end_label)

File: app/src/main/java/com/ewaldmire/osmride/ui/bodymetrics/TrendChart.kt
package com.ewaldmire.osmride.ui.bodymetrics

import android.graphics.Paint
import androidx.compose.foundation.Canvas
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.sp
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

private val axisDateFormatter = DateTimeFormatter.ofPattern("MMM d")
private const val ROLLING_WINDOW_MILLIS = 7L * 24 * 60 * 60 * 1000

/** A single measurement in a trend series - used for both weight and waist, since the chart
 * itself doesn't care what unit the value is in (callers pre-convert for display). */
data class TrendPoint(val epochMillis: Long, val value: Double)

/**
 * Value-vs-time strip: raw daily points as a thin faded line/dots, with a bold 7-day rolling
 * average line and filled area on top - daily weight/waist fluctuates from water, sodium, timing,
 * etc., so the average is the actionable trend, but showing the raw noise too keeps a bad single
 * day from misreading as a real regression.
 */
@Composable
fun TrendChart(points: List<TrendPoint>, modifier: Modifier = Modifier) {
    val chronological = remember(points) { points.sortedBy { it.epochMillis } }
    val averaged = remember(chronological) { rollingAverage(chronological, ROLLING_WINDOW_MILLIS) }

    val allValues = chronological.map { it.value } + averaged.map { it.value }
    val minValue = remember(allValues) { allValues.minOrNull() ?: 0.0 }
    val maxValue = remember(allValues) { (allValues.maxOrNull() ?: 1.0).coerceAtLeast(minValue + 0.1) }
    val minTime = chronological.firstOrNull()?.epochMillis ?: 0L
    val maxTime = (chronological.lastOrNull()?.epochMillis ?: (minTime + 1L)).coerceAtLeast(minTime + 1L)

    val backgroundColor = MaterialTheme.colorScheme.surfaceVariant
    val fillColor = MaterialTheme.colorScheme.primary
    val rawColor = MaterialTheme.colorScheme.onSurfaceVariant
    val labelColor = MaterialTheme.colorScheme.onSurface
    val labelSizePx = with(LocalDensity.current) { 12.sp.toPx() }

    val startLabel = remember(minTime) {
        Instant.ofEpochMilli(minTime).atZone(ZoneId.systemDefault()).format(axisDateFormatter)
    }
    val endLabel = remember(maxTime) {
        Instant.ofEpochMilli(maxTime).atZone(ZoneId.systemDefault()).format(axisDateFormatter)
    }

    Canvas(modifier = modifier) {
        drawRect(color = backgroundColor)

        if (chronological.size < 2) return@Canvas

        fun xFor(epochMillis: Long) = ((epochMillis - minTime).toFloat() / (maxTime - minTime).toFloat()) * size.width
        fun yFor(value: Double) = size.height - ((value - minValue) / (maxValue - minValue)).toFloat() * size.height

        // Raw daily points - thin, faded, no fill, so they read as background noise not signal.
        val rawPath = Path().apply {
            chronological.forEachIndexed { index, point ->
                val x = xFor(point.epochMillis)
                val y = yFor(point.value)
                if (index == 0) moveTo(x, y) else lineTo(x, y)
            }
        }
        drawPath(rawPath, color = rawColor.copy(alpha = 0.35f), style = Stroke(width = 2f))
        chronological.forEach { point ->
            drawCircle(
                color = rawColor.copy(alpha = 0.35f),
                radius = 3f,
                center = Offset(xFor(point.epochMillis), yFor(point.value)),
            )
        }

        // 7-day rolling average - the actual signal, drawn bold with a filled area underneath.
        val avgPath = Path().apply {
            moveTo(0f, size.height)
            averaged.forEach { point -> lineTo(xFor(point.epochMillis), yFor(point.value)) }
            lineTo(size.width, size.height)
            close()
        }
        drawPath(avgPath, color = fillColor.copy(alpha = 0.5f))
        val avgLine = Path().apply {
            averaged.forEachIndexed { index, point ->
                val x = xFor(point.epochMillis)
                val y = yFor(point.value)
                if (index == 0) moveTo(x, y) else lineTo(x, y)
            }
        }
        drawPath(avgLine, color = fillColor, style = Stroke(width = 5f))

        // Without these, the line's x-position was scaled by real elapsed time but nothing on
        // screen actually told the viewer that - it just looked like an arbitrary shape.
        drawIntoCanvas { canvas ->
            val paint = Paint().apply {
                color = labelColor.toArgb()
                textSize = labelSizePx
                isAntiAlias = true
            }
            canvas.nativeCanvas.drawText(startLabel, 4f, size.height - 6f, paint)
            paint.textAlign = Paint.Align.RIGHT
            canvas.nativeCanvas.drawText(endLabel, size.width - 4f, size.height - 6f, paint)
        }
    }
}

/** points must be chronological (oldest first). Each output point is the average of every input
 * point within [windowMillis] before (and including) it - a trailing average, not centered, so
 * it only ever needs data that's already been logged. */
private fun rollingAverage(points: List<TrendPoint>, windowMillis: Long): List<TrendPoint> {
    val result = ArrayList<TrendPoint>(points.size)
    var windowStartIndex = 0
    var windowSum = 0.0
    for (i in points.indices) {
        windowSum += points[i].value
        val cutoff = points[i].epochMillis - windowMillis
        while (points[windowStartIndex].epochMillis <= cutoff) {
            windowSum -= points[windowStartIndex].value
            windowStartIndex++
        }
        val windowCount = i - windowStartIndex + 1
        result.add(TrendPoint(points[i].epochMillis, windowSum / windowCount))
    }
    return result
}
